app/controllers/application_controller.rb

```diff
@@ -5,6 +5,151 @@ class ApplicationController < Sinatra::Base
   configure do
     set :public_folder, 'public'
     set :views, 'app/views'
+    enable :sessions
+    set :session_secret, "password_security"
   end
 
-end
\ No newline at end of file
+  get '/tweets/new' do
+    if logged_in?
+      erb :'tweets/create_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets' do
+    if logged_in?
+      if !params[:content].empty?
+        @tweet = Tweet.create(params)
+        current_user.tweets << @tweet
+        redirect "/tweets/#{@tweet.id}"
+      else
+        redirect '/tweets/new'
+      end
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/'do
+    if logged_in?
+      redirect '/tweets'
+    else
+      erb :index
+    end
+  end
+
+  get '/signup' do
+    if logged_in?
+      redirect '/tweets'
+    else
+      erb :'users/create_user'
+    end
+  end
+
+  post '/signup' do
+    if !params[:username].empty? && !params[:email].empty? && !params[:password].empty?
+      @user = User.create(params)
+      session[:user_id] = @user.id
+      redirect '/tweets'
+    else
+      redirect '/signup'
+    end
+  end
+
+  get '/login' do
+    if logged_in?
+      redirect '/tweets'
+    else
+      erb :'users/login'
+    end
+  end
+
+  post '/login' do
+    user = User.find_by(:username => params[:username])
+    if user && user.authenticate(params[:password])
+      session[:user_id] = user.id
+      redirect '/tweets'
+    elsif user
+      redirect '/login'
+    else
+      redirect '/signup'
+    end
+
+  end
+
+  get '/tweets' do
+    if logged_in?
+      erb :'tweets/tweets'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/logout' do
+    if logged_in?
+      session.clear
+    end
+    redirect '/login'
+  end
+
+  get '/tweets/:id' do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      erb :'tweets/show_tweet'
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/tweets/:id/edit' do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      if current_user.id == @tweet.user.id
+        erb :'tweets/edit_tweet'
+      else
+        redirect "/tweets/#{@tweet.id}"
+      end
+    else
+      redirect '/login'
+    end
+  end
+
+  post '/tweets/:id' do
+    @tweet = Tweet.find(params[:id])
+    if logged_in?
+      if !params[:content].empty?
+        @tweet.update(params) if current_user.id == @tweet.user.id
+      end
+      redirect "/tweets/#{@tweet.id}/edit"
+    else
+      redirect '/login'
+    end
+  end
+
+  use Rack::MethodOverride
+  delete '/tweets/:id/delete' do
+    if logged_in?
+      @tweet = Tweet.find(params[:id])
+      @tweet.destroy if current_user.id == @tweet.user.id
+      redirect "/tweets"
+    else
+      redirect '/login'
+    end
+  end
+
+  get '/users/:slug' do
+    @user = User.find_by_slug(params[:slug])
+    erb :'users/show'
+  end
+
+  helpers do
+    def current_user
+      User.find(session[:user_id])
+    end
+
+    def logged_in?
+      !!session[:user_id]
```

Comment:
It's better to rely on the successful call to `current_user` than only introspecting on the `session[:id]` value. In theory, a user might have an invalid `id` in their session which would make `logged_in?` return true even though that user would not be findable later. Also, because `current_user` is memoized, it makes sense to call it as early in the request cycle as possible, which generally is the first call `logged_in?`.

```ruby
def logged_in?
  !!current_user
end
```